Track global episode events - Part 2

ashiagr opened the pull request:
## Description
This tracks below episode events:
- `episode_starred`: When an episode is starred
- `episode_bulk_starred`: When the user stars multiple episodes while in multi select mode
- `episode_unstarred`: When the user stars an episode
- `episode_bulk_unstarred`: "When the user unstars multiple episodes while in multi select mode
- `episode_archived`: When the user archives an episode
- `episode_bulk_archived`: "When the user archives multiple episodes while in multi select mode
- `episode_unarchived`: "When the user unarchives an episode
- `episode_bulk_unarchived`: "When the user unarchives multiple episodes while in multi select mode
- `episode_marked_as_played`: When the user marks an episode as played
- `episode_bulk_marked_as_played`: "When the user marks multiple episodes as played while in multi select mode
- `episode_marked_as_unplayed`: When the user marks the episode as unplayed
- `episode_bulk_marked_as_unplayed`: "When the user marks multiple episodes as unplayed while in multi select mode
- `episode_added_to_up_next`: When the user adds an episode to their up next
- `episode_bulk_add_to_up_next`: "When the user adds multiple episodes to the up next queue while in multi select mode
- `episode_removed_from_up_next`: When the user removes an episode from their up next

iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/298
Part 1 PR: https://github.com/Automattic/pocket-casts-android/pull/684

## Testing Instructions
<details><summary>Star / Unstar</summary>

1. Tap on the Podcasts tab
2. Tap on a podcast
3. Tap on an episode
4. Tap the Star icon 
5. ✅ `🔵 Tracked: episode_starred ["source": "episode_detail", "episode_uuid": "UUID"]`
6. Tap the star icon again
7. ✅ `🔵 Tracked: episode_unstarred ["episode_uuid": "UUID", "source": "episode_detail"]`
8. Dismiss the episode details
9. Tap and hold on the podcast list to enter multi select mode
10. Select multiple episodes
11. Tap the ... if the star action is not in your shelf
12. Tap the Star Episodes action
13. ✅ `🔵 Tracked: episode_bulk_starred ["episode_count": COUNT, "source": "podcast_screen"]` - Verify the count is the number of episodes you selected
14. Enter multi select mode again
15. Select the same episodes
16. Tap the Unstar Episodes action
17. ✅ `🔵 Tracked: episode_bulk_unstarred ["episode_count": COUNT, "source": "podcast_screen"]`

**Repeat the steps on the following views:**
1. Filter Details
2. Downloads
3. Starred
4. Listening History

**Player Starred**
1. Open the full screen player
2. Tap the Star action in the shelf
3. ✅  Verify you see: `🔵 Tracked: episode_starred ["source": "player", "episode_uuid": "UUID"]`
4. Unstar the episode
5. ✅ `🔵 Tracked: episode_unstarred ["episode_uuid": "UUID", "source": "player"]`

</details>

<details><summary>Archived / Unarchived</summary>

1. Tap on the Podcasts tab
2. Tap on a podcast
3. Tap on an episode
4. Tap the Archive button
5. ✅ `🔵 Tracked: episode_archived ["source": "episode_detail", "episode_uuid": "UUID"]`
6. Locate the episode in the podcast list again and tap it
7. Tap the Unarchive button
8. ✅ `🔵 Tracked: episode_unarchived ["episode_uuid": "UUID", "source": "episode_detail"]`
9. Dismiss the episode details
10. Swipe to the left to expose the right action
11. Tap it
12. ✅ `🔵 Tracked: episode_archived ["episode_uuid": "UUID", "source": "podcast_screen"]`
13. Tap the swipe action again
14. ✅ `🔵 Tracked: episode_unarchived ["episode_uuid": "UUID", "source": "podcast_screen"]`
15. Tap and hold on the podcast list to enter multi select mode
16. Select multiple episodes
17. Tap the ... if the archive action is not in your shelf
18. Tap the Archive Episodes action
19. ✅ `🔵 Tracked: episode_bulk_archived ["episode_count": COUNT, "source": "podcast_screen"]` - Verify the count is the number of episodes you selected
20. Enter multi select mode again
21. Select the same episodes
22. Tap the Unarchive Episodes action
23. ✅ `🔵 Tracked: episode_bulk_unarchived ["episode_count": COUNT, "source": "podcast_screen"]`

**Repeat the steps on the following views:**
1. Filter Details
2. Downloads
3. Starred
4. Listening History

**Player Archive**
1. Open the full screen player
2. Tap the Archive action in the shelf
3. ✅  Verify you see: `🔵 Tracked: episode_archived ["source": "player", "episode_uuid": "UUID"]`
</details>

<details><summary>Mark as Played / Unplayed</summary>

1. Tap on the Podcasts tab
2. Tap on a podcast
3. Tap on an episode
4. Tap the Mark Played button
5. ✅ `🔵 Tracked: episode_marked_as_played ["source": "episode_detail", "episode_uuid": "UUID"]`
7. Tap the Mark Unplayed button
8. ✅ `🔵 Tracked: episode_marked_as_unplayed ["episode_uuid": "UUID", "source": "episode_detail"]`
9. Dismiss the episode details
10. Tap and hold on the podcast list to enter multi select mode
11. Select multiple episodes
12. Tap the ... if the Mark as Played action is not in your shelf
13. Tap the Mark as Played action
14. ✅ `🔵 Tracked: episode_bulk_marked_as_played ["episode_count": COUNT, "source": "podcast_screen"]` - Verify the count is the number of episodes you selected
15. Enter multi select mode again
16. Select the same episodes
17. Tap the Mark as Unplayed action
18. ✅ `🔵 Tracked: episode_bulk_marked_as_unplayed ["episode_count": COUNT, "source": "podcast_screen"]`

**Repeat the steps on the following views:**
1. Filter Details
2. Downloads
3. Starred
4. Listening History
5. Files

**Player Mark as Played**
1. Open the full screen player
2. Tap the Mark as Played action in the shelf
3. ✅  Verify you see: `🔵 Tracked: episode_marked_as_played ["source": "player", "episode_uuid": "UUID"]`
</details>

<details><summary>Add to up next / Remove from up next</summary>

1. Tap on the Podcasts tab
2. Tap on a podcast
3. Tap on an episode
4. Tap the Add Up next button
6. ✅ `🔵 Tracked: episode_added_to_up_next ["to_top": false, "source": "episode_detail", "episode_uuid": "UUID"]`
7. Tap the Remove from Up Next button
8. ✅ `🔵 Tracked: episode_removed_from_up_next ["episode_uuid": "UUID", "source": "episode_detail"]`
9. Dismiss the episode details
10. Swipe right to expose the left action
11. Tap the Add to the top action
12. ✅ `🔵 Tracked: episode_added_to_up_next ["episode_uuid": "UUID", "to_top": true, "source": "podcast_screen"]`
13. Swipe again, and select the remove item
14. ✅ `🔵 Tracked: episode_removed_from_up_next ["source": "podcast_screen", "episode_uuid": "UUID"]`
15. Swipe again and select the add to bottom item
16. ✅ `🔵 Tracked: episode_added_to_up_next ["episode_uuid": "UUID", "to_top": false, "source": "podcast_screen"]`
17. Tap and hold on the podcast list to enter multi select mode
18. Select multiple episodes
24. Tap the ... if the Play Next/Last actions are not in your shelf
25. Tap the Play Next action
26. ✅ `🔵 Tracked: episode_bulk_add_to_up_next ["source": "podcast_screen", "episode_count": COUNT, "to_top": true]` - Verify the count is the number of episodes you selected
27. Enter multi select mode again
28. Select the same episodes
29. Tap the Play Last action
30. ✅ `🔵 Tracked: episode_bulk_add_to_up_next ["source": "podcast_screen", "episode_count": COUNT, "to_top": false]` - Verify the count is the number of episodes you selected

**Repeat the steps on the following views:**
1. Filter Details
2. Downloads
3. Starred
4. Listening History
5. Files

</details>


This also tracks failed upload and download events in c965015 as suggested in https://github.com/Automattic/pocket-casts-android/pull/684#pullrequestreview-1242478195.

mchowning reviewed (approved):
This is looking good. Thanks for knocking all of these out!

The only issue I see is that I only got the `episode_archived` / `episode_unarchived` event for me when swiping episodes on the podcast screen. I am not getting those events (I would only get the `episode_swipe_action_perfomed` event) from the Filters, Starred, Downloads, and Listening History screens.

Also looks like some tests are still failing.

ashiagr commented:
> The only issue I see is that I only got the episode_archived / episode_unarchived event for me when swiping episodes on the podcast screen. I am not getting those events (I would only get the episode_swipe_action_perfomed event) from the Filters, Starred, Downloads, and Listening History screens.

Thanks for catching it. 👍 Added in fe1605c. 

> Also looks like some tests are still failing.

Fixed in 4aacb7e. I moved star/ unstar tracking from `EpisodeManagerImpl` to source classes and removed the problematic analytics mock in order to fix it.

mchowning reviewed (approved):
Looks good! 👍
